r/src/array__to_vector.cpp

```diff
@@ -0,0 +1,827 @@
+      k += n_chunk;
+    }
+    return Status::OK();
+  }
+};
+
+struct CanParallel {
+  using OUT = bool;
+
+  template <typename Converter>
+  static bool Do(const ArrayVector& arrays) {
+    return parallel_ingest<Converter>();
+  }
+};
+
+// Only allocate an R vector to host the arrays
+SEXP ArrayVector__Allocate(R_xlen_t n, const ArrayVector& arrays) {
+  return ArrayVector__Dispatch<arrow::r::Allocator, R_xlen_t>(arrays, n);
+}
+
+// Ingest data from arrays to previously allocated R vector
+// For most vector types, this can be done in a task
+// in some other thread
+Status ArrayVector__Ingest(SEXP data, const ArrayVector& arrays) {
+  return ArrayVector__Dispatch<arrow::r::Ingester, SEXP>(arrays, data);
+}
+
+bool ArrayVector__Parallel(const ArrayVector& arrays) {
+  return ArrayVector__Dispatch<arrow::r::CanParallel>(arrays);
+}
+
+// Allocate + Ingest
+SEXP ArrayVector__as_vector(R_xlen_t n, const ArrayVector& arrays) {
+  Shield<SEXP> data(ArrayVector__Allocate(n, arrays));
+  STOP_IF_NOT_OK(ArrayVector__Ingest(data, arrays));
+  return data;
+}
+
+}  // namespace r
+}  // namespace arrow
+
+// [[Rcpp::export]]
+SEXP Array__as_vector(const std::shared_ptr<arrow::Array>& array) {
+  return arrow::r::ArrayVector__as_vector(array->length(), {array});
+}
+
+// [[Rcpp::export]]
+SEXP ChunkedArray__as_vector(const std::shared_ptr<arrow::ChunkedArray>& chunked_array) {
+  return arrow::r::ArrayVector__as_vector(chunked_array->length(),
+                                          chunked_array->chunks());
+}
+
+List RecordBatch__to_dataframe_serial(const std::shared_ptr<arrow::RecordBatch>& batch) {
+  auto nc = batch->num_columns();
+  auto nr = batch->num_rows();
+  List tbl(nc);
+  CharacterVector names(nc);
+
+  for (int i = 0; i < nc; i++) {
+    tbl[i] = Array__as_vector(batch->column(i));
+    names[i] = batch->column_name(i);
+  }
+  tbl.attr("names") = names;
+  tbl.attr("class") = CharacterVector::create("tbl_df", "tbl", "data.frame");
+  tbl.attr("row.names") = IntegerVector::create(NA_INTEGER, -nr);
+  return tbl;
+}
+
+List RecordBatch__to_dataframe_parallel(
+    const std::shared_ptr<arrow::RecordBatch>& batch) {
+  auto nc = batch->num_columns();
+  auto nr = batch->num_rows();
+  List tbl(nc);
+  CharacterVector names(nc);
+
+  // task group to ingest data in parallel
+  auto tg = arrow::internal::TaskGroup::MakeThreaded(arrow::internal::GetCpuThreadPool());
+
+  // allocate and start ingesting immediately the columns that
+  // can be ingested in parallel, i.e. when ingestion no longer
+  // need to happen on the main thread
+  for (int i = 0; i < nc; i++) {
+    ArrayVector arrays{batch->column(i)};
+    // allocate data for column i
+    SEXP column = tbl[i] = arrow::r::ArrayVector__Allocate(nr, arrays);
+
+    // add a task to ingest data of that column if that can be done in parallel
+    if (arrow::r::ArrayVector__Parallel(arrays)) {
+      tg->Append([=] { return arrow::r::ArrayVector__Ingest(column, arrays); });
```

Comment:
I have not looked at ingesting the chunks of a chunk array in parallel, i.e. get one task per chunk instead of one task per ChunkedArray. 

I suppose it could make a difference if you have a Table that has 2 columns but many chunks for each.

Reply:
Yep, I just thought of it as I was looking at this code since we have not employed the task groups at all set in the Python converter layer

Reply:
are tasks cheap ?

Reply:
Yes, I believe so. It would be useful to have benchmarks to quantify

https://issues.apache.org/jira/browse/ARROW-4177

Reply:
You should definitely ingest individual chunks, IMHO.

Reply:
fair enough. Will do.